File: new/caller_bot/exotel_ivr.py
from fastapi import FastAPI, Request, Form, HTTPException
from fastapi.responses import PlainTextResponse
import httpx
import logging
import os
import uvicorn
from dotenv import load_dotenv

# Load environment variables
load_dotenv()

# Setup logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

# Initialize FastAPI app
app = FastAPI(title="Exotel IVR System")

# Define the base URLs for APIs
TRANSLATOR_API = os.getenv("TRANSLATOR_API", "http://localhost:8000")
RAG_API_URL = os.getenv("RAG_API_URL", "http://localhost:8000") # RAG API URL

# Define answers for each option
answers = {
    1: "For Right to Information Act, you need to file an RTI application with the concerned Public Information Officer. The fee is generally Rs. 10, and you can expect a response within 30 days.",
    2: "Under Indian Penal Code, Section 420 deals with cheating and dishonestly inducing delivery of property, punishable with imprisonment up to 7 years and fine.",
    3: "For labor law matters, you can approach the labor commissioner or file a case in the labor court. You may also seek help from trade unions if available in your workplace.",
    4: "As per the Constitution of India, Article 32 provides the right to constitutional remedies, which allows citizens to move the Supreme Court for enforcement of fundamental rights."
}

# Exotel AppID and Secret for authentication
EXOTEL_APP_ID = os.getenv("EXOTEL_APP_ID")
EXOTEL_APP_SECRET = os.getenv("EXOTEL_APP_SECRET")

# Map numeric options to bot names
bot_mapping = {
    1: "RTI Bot",
    2: "IPC Bot",
    3: "Labor Law Bot",
    4: "Constitution Bot"
}

# Store user context
user_sessions = {}

# Generate welcome menu in various languages
async def generate_menu_audios():
    """Generate welcome and menu audio files in supported languages"""
    lang_clients = httpx.AsyncClient()
    
    welcome_text = {
        "english": "Welcome to the Legal Assistant. Press 1 for English, 2 for Hindi.",
        "hindi": "कानूनी सहायक में आपका स्वागत है। अंग्रेजी के लिए 1 दबाएं, हिंदी के लिए 2 दबाएं।"
    }
    
    menu_text = {
        "english": "Please select an option. Press 1 for RTI information. Press 2 for IPC information. Press 3 for labor laws. Press 4 for constitutional rights.",
        "hindi": "कृपया एक विकल्प चुनें। आरटीआई जानकारी के लिए 1 दबाएं। आईपीसी जानकारी के लिए 2 दबाएं। श्रम कानूनों के लिए 3 दबाएं। संवैधानिक अधिकारों के लिए 4 दबाएं।"
    }
    
    # Map to store audio file paths
    audio_files = {}
    
    for lang, text in welcome_text.items():
        try:
            # Create welcome audio
            response = await lang_clients.post(
                f"{TRANSLATOR_API}/translate/{lang}",
                json={"text": text}
            )
            response.raise_for_status()
            audio_files[f"welcome_{lang}"] = response.json()["audio_file"]
            
            # Create menu audio
            response = await lang_clients.post(
                f"{TRANSLATOR_API}/translate/{lang}",
                json={"text": menu_text[lang]}
            )
            response.raise_for_status()
            audio_files[f"menu_{lang}"] = response.json()["audio_file"]
            
            logger.info(f"Created menu audios for {lang}")
        except Exception as e:
            logger.error(f"Failed to create {lang} audios: {e}")
    
    await lang_clients.aclose()
    return audio_files

# Initialize audio files - will be populated during startup
menu_audios = {}

@app.on_event("startup")
async def startup_event():
    """Generate menu audios on startup"""
    global menu_audios
    try:
        menu_audios = await generate_menu_audios()
        logger.info("Generated menu audio files")
    except Exception as e:
        logger.error(f"Failed to generate menu audios: {e}")

@app.get("/")
def read_root():
    return {"message": "Exotel IVR System is running"}

@app.post("/ivr/welcome")
async def welcome_ivr(
    CallSid: str = Form(...),
    From: str = Form(...),
    digits_entered: str = Form(None)
):
    """Initial welcome message and language selection"""
    logger.info(f"Call from {From}, CallSid: {CallSid}, digits: {digits_entered}")
    
    # If no digits entered, play welcome message
    if not digits_entered:
        # Construct Exotel OBD flow XML to play welcome and gather digits
        welcome_audio = menu_audios.get("welcome_english", "")
        
        response_xml = f"""<?xml version="1.0" encoding="UTF-8"?>
        <Response>
            <Play>{TRANSLATOR_API}/audio/{welcome_audio}</Play>
            <GetDigits timeout="5" numDigits="1" callbackUrl="{request.url}"/>
        </Response>
        """
        return PlainTextResponse(content=response_xml, media_type="application/xml")
    
    # Process language selection
    lang = "english" if digits_entered == "1" else "hindi" if digits_entered == "2" else "english"
    
    # Play menu options in selected language
    menu_audio = menu_audios.get(f"menu_{lang}", "")
    
    response_xml = f"""<?xml version="1.0" encoding="UTF-8"?>
    <Response>
        <Play>{TRANSLATOR_API}/audio/{menu_audio}</Play>
        <GetDigits timeout="5" numDigits="1" callbackUrl="/ivr/menu/{lang}"/>
    </Response>
    """
    return PlainTextResponse(content=response_xml, media_type="application/xml")

@app.post("/ivr/menu/{language}")
async def menu_selection(
    language: str,
    CallSid: str = Form(...),
    From: str = Form(...),
    digits_entered: str = Form(None)
):
    """Process menu selection and play appropriate answer"""
    logger.info(f"Menu selection for {language}, CallSid: {CallSid}, Digits: {digits_entered}")
    
    # Validate input
    if not digits_entered or digits_entered not in ["1", "2", "3", "4"]:
        # Invalid or no input, replay menu
        menu_audio = menu_audios.get(f"menu_{language}", "")
        
        response_xml = f"""<?xml version="1.0" encoding="UTF-8"?>
        <Response>
            <Say>Invalid selection.</Say>
            <Play>{TRANSLATOR_API}/audio/{menu_audio}</Play>
            <GetDigits timeout="5" numDigits="1" callbackUrl="/ivr/menu/{language}"/>
        </Response>
        """
        return PlainTextResponse(content=response_xml, media_type="application/xml")
    
    # Store the selected bot in the session
    option_num = int(digits_entered)
    selected_bot = bot_mapping.get(option_num)
    
    user_sessions[CallSid] = {
        "selected_bot": selected_bot,
        "language": language
    }
    
    # Prompt user to ask a question
    prompt_text = {
        "english": f"You've selected {selected_bot}. Please ask your question after the beep.",
        "hindi": f"आपने {selected_bot} का चयन किया है। कृपया बीप के बाद अपना सवाल पूछें।"
    }.get(language, f"You've selected {selected_bot}. Please ask your question after the beep.")

    try:
        # Get translated prompt
        async with httpx.AsyncClient() as client:
            response = await client.post(
                f"{TRANSLATOR_API}/translate/{language}",
                json={"text": prompt_text}
            )
            response.raise_for_status()
            prompt_data = response.json()
            audio_file = prompt_data["audio_file"]
            
            # Ask user to record their question
            response_xml = f"""<?xml version="1.0" encoding="UTF-8"?>
            <Response>
                <Play>{TRANSLATOR_API}/audio/{audio_file}</Play>
                <Record maxLength="30" playBeep="true" callbackUrl="/ivr/process_question/{CallSid}"/>
            </Response>
            """
            return PlainTextResponse(content=response_xml, media_type="application/xml")
    except Exception as e:
        logger.error(f"Error creating prompt: {e}")
        
        # Fallback response
        response_xml = """<?xml version="1.0" encoding="UTF-8"?>
        <Response>
            <Say>Sorry, we couldn't process your request. Please try again later.</Say>
            <Hangup/>
        </Response>
        """
        return PlainTextResponse(content=response_xml, media_type="application/xml")

@app.post("/ivr/process_question/{call_sid}")
async def process_question(
    call_sid: str,
    CallSid: str = Form(...),
    From: str = Form(...),
    RecordingUrl: str = Form(None),
    RecordingStatus: str = Form(None)
):
    """Process the recorded question and get answer from RAG"""
    logger.info(f"Processing question for Call: {call_sid}, Recording: {RecordingUrl}, Status: {RecordingStatus}")
    
    if RecordingStatus != "completed" or not RecordingUrl:
        response_xml = """<?xml version="1.0" encoding="UTF-8"?>
        <Response>
            <Say>We couldn't record your question. Please try again later.</Say>
            <Hangup/>
        </Response>
        """
        return PlainTextResponse(content=response_xml, media_type="application/xml")
    
    try:
        # Get session data
        session = user_sessions.get(call_sid, {})
        selected_bot = session.get("selected_bot")
        language = session.get("language", "english")
        
        if not selected_bot:
            raise ValueError("No bot selected for this call")
        
        # 1. Convert speech to text
        async with httpx.AsyncClient() as client:
            # Send recording to speech-to-text service
            stt_response = await client.post(
                f"{RAG_API_URL}/speech-to-text",
                json={"audio_url": RecordingUrl, "language": language}
            )
            stt_response.raise_for_status()
            stt_data = stt_response.json()
            question_text = stt_data["text"]
            
            logger.info(f"Transcribed question: {question_text}")
            
            # 2. Query RAG API with the question
            rag_response = await client.post(
                f"{RAG_API_URL}/bots/{selected_bot}/query",
                json={"query": question_text}
            )
            rag_response.raise_for_status()
            rag_data = rag_response.json()
            answer_text = rag_data["answer"]
            
            # 3. Convert answer to speech in the selected language
            tts_response = await client.post(
                f"{RAG_API_URL}/text-to-speech",
                json={"text": answer_text, "language": language}
            )
            tts_response.raise_for_status()
            tts_data = tts_response.json()
            audio_file = tts_data["audio_file"]
            
            # 4. Play the answer to the user with option to ask another question
            response_xml = f"""<?xml version="1.0" encoding="UTF-8"?>
            <Response>
                <Play>{RAG_API_URL}/audio/{audio_file}</Play>
                <Gather numDigits="1" timeout="5" action="/ivr/after_answer/{call_sid}">
                    <Say>Press 1 to ask another question. Press 2 to end the call.</Say>
                </Gather>
                <Say>Thank you for using our service. Goodbye.</Say>
                <Hangup/>
            </Response>
            """
            return PlainTextResponse(content=response_xml, media_type="application/xml")
    except Exception as e:
        logger.error(f"Error processing question: {e}")
        
        # Fallback response
        response_xml = """<?xml version="1.0" encoding="UTF-8"?>
        <Response>
            <Say>Sorry, we couldn't process your question. Please try again later.</Say>
            <Hangup/>
        </Response>
        """
        return PlainTextResponse(content=response_xml, media_type="application/xml")

@app.post("/ivr/after_answer/{call_sid}")
async def after_answer(
    call_sid: str,
    CallSid: str = Form(...),
    From: str = Form(...),
    Digits: str = Form(None)
):
    """Handle user's choice after hearing the answer"""
    logger.info(f"After answer for Call: {call_sid}, Choice: {Digits}")
    
    session = user_sessions.get(call_sid, {})
    selected_bot = session.get("selected_bot")
    language = session.get("language", "english")
    
    if Digits == "1" and selected_bot:
        # User wants to ask another question
        prompt_text = {
            "english": "Please ask your next question after the beep.",
            "hindi": "कृपया बीप के बाद अपना अगला सवाल पूछें।"
        }.get(language, "Please ask your next question after the beep.")
        
        try:
            async with httpx.AsyncClient() as client:
                response = await client.post(
                    f"{TRANSLATOR_API}/translate/{language}",
                    json={"text": prompt_text}
                )
                response.raise_for_status()
                prompt_data = response.json()
                audio_file = prompt_data["audio_file"]
                
                # Ask user to record their next question
                response_xml = f"""<?xml version="1.0" encoding="UTF-8"?>
                <Response>
                    <Play>{TRANSLATOR_API}/audio/{audio_file}</Play>
                    <Record maxLength="30" playBeep="true" callbackUrl="/ivr/process_question/{call_sid}"/>
                </Response>
                """
                return PlainTextResponse(content=response_xml, media_type="application/xml")
        except Exception as e:
            logger.error(f"Error creating prompt: {e}")
    
    # Default: end call
    goodbye_text = {
        "english": "Thank you for using our Legal Assistant. Goodbye.",
        "hindi": "हमारे लीगल असिस्टेंट का उपयोग करने के लिए धन्यवाद। अलविदा।"
    }.get(language, "Thank you for using our Legal Assistant. Goodbye.")
    
    response_xml = f"""<?xml version="1.0" encoding="UTF-8"?>
    <Response>
        <Say>{goodbye_text}</Say>
        <Hangup/>
    </Response>
    """
    return PlainTextResponse(content=response_xml, media_type="application/xml")

@app.get("/health")
async def health_check():
    """Health check endpoint"""
    try:
        # Check if RAG API is accessible
        async with httpx.AsyncClient() as client:
            rag_health = await client.get(f"{RAG_API_URL}/health", timeout=2.0)
            rag_health.raise_for_status()
            
        return {
            "status": "healthy",
            "rag_api": "connected",
            "sessions_active": len(user_sessions)
        }
    except Exception as e:
        logger.error(f"Health check failed: {e}")
        return {
            "status": "degraded",
            "error": str(e),
            "sessions_active": len(user_sessions)
        }

if __name__ == "__main__":
    uvicorn.run("exotel_ivr:app", host="0.0.0.0", port=8080, reload=True)

File: new/callerjs/exotel-ivr.js
require('dotenv').config();
const express = require('express');
const axios = require('axios');
const { GoogleGenerativeAI } = require('@google/generative-ai');
const app = express();
const PORT = process.env.PORT || 3009;

// Initialize Google Gemini API
const genAI = new GoogleGenerativeAI(process.env.GOOGLE_API_KEY);
const model = genAI.getGenerativeModel({ model: "gemini-2.0-flash" });

// Configure RAG API (your existing API with the legal bots)
const RAG_API_URL = process.env.RAG_API_URL || 'http://localhost:8000';

// Track call sessions
const callSessions = {};

// Available legal bots with enhanced descriptions
const LEGAL_BOTS = {
  '1': {
    name: 'RTI Bot',
    shortDesc: 'Right to Information',
    response: 'For Right to Information Act requests, you need to file an application with a fee of 10 rupees. Public authorities must respond within 30 days. You can appeal if denied information within 30 days to the first appellate authority.'
  },
  '2': {
    name: 'IPC Bot',
    shortDesc: 'Indian Penal Code',
    response: 'Under Indian Penal Code Section 420, cheating is punishable with imprisonment up to 7 years and a fine. For other sections like 302 for murder or 376 for rape, punishments vary accordingly.'
  },
  '3': {
    name: 'Labor Law Bot',
    shortDesc: 'Labor Laws',
    response: 'For labor disputes, approach your local labor commissioner first. Employees are entitled to minimum wages, overtime pay, and safe working conditions as per the labor laws.'
  },
  '4': {
    name: 'Constitution Bot',
    shortDesc: 'Constitutional Rights',
    response: 'Article 32 of the Constitution provides right to constitutional remedies. This fundamental right allows citizens to approach the Supreme Court directly for enforcement of their rights.'
  },
  '5': {
    name: 'Family Law Bot',
    shortDesc: 'Family & Marriage Laws',
    response: 'Family law covers marriage, divorce, adoption, and child custody. For divorce, different personal laws have different waiting periods and grounds depending on your religion.'
  }
};

// Parse form data (Exotel sends form-urlencoded data)
app.use(express.urlencoded({ extended: true }));
app.use(express.json());

// Log incoming requests with more detail
app.use((req, res, next) => {
  console.log(`[${new Date().toISOString()}] ${req.method} ${req.path}`);
  console.log('Request query:', req.query);
  if (Object.keys(req.body).length > 0) {
    console.log('Request body:', req.body);
  }
  
  // Add response logging
  const originalSend = res.send;
  res.send = function(body) {
    console.log('Response sent:', body);
    return originalSend.call(this, body);
  };
  
  next();
});

// Set BASE_URL for callbacks - make sure to use a direct URL without dynamic components
const getBaseUrl = (req) => {
  return process.env.BASE_URL || `${req.protocol}://${req.get('host')}`;
};

// ROOT ENDPOINT - Modified to use Play tag for audio URLs instead of Say tag
app.get('/', (req, res) => {
  try {
    // Extract Exotel call parameters
    const callSid = req.query.CallSid || 'unknown';
    const callerNumber = req.query.CallFrom || req.query.From || '0000000000';
    
    console.log(`New call from ${callerNumber}, CallSid: ${callSid}`);
    
    // Initialize session for this call
    callSessions[callSid] = {
      callerNumber,
      stage: 'bot-selection',  // Skip greeting since Exotel already did it
      language: 'en',
      selectedBot: null,
      lastQuery: null,
      lastResponse: null,
      timestamp: new Date().toISOString()
    };
    
    // Using Play tag with audio URLs instead of Say tag
    const exotelResponse = `<?xml version="1.0" encoding="UTF-8"?>
<Response>
<Play>${process.env.AUDIO_BASE_URL}/menu-options.mp3</Play>
<Gather action="${getBaseUrl(req)}/bot-selection" numDigits="1"/>
</Response>`;
    
    console.log('Sending domain selection options with audio');
    res.set('Content-Type', 'text/xml');
    return res.send(exotelResponse);
  } catch (error) {
    console.error('Error in root handler:', error);
    const errorResponse = `<?xml version="1.0" encoding="UTF-8"?>
<Response>
<Play>${process.env.AUDIO_BASE_URL}/error.mp3</Play>
<Hangup/>
</Response>`;
    res.set('Content-Type', 'text/xml');
    return res.send(errorResponse);
  }
});

// Handle both GET and POST for bot selection with simplified XML response
app.all('/bot-selection', (req, res) => {
  try {
    // Get parameters from either query or body
    const params = {...req.query, ...req.body};
    console.log('Bot selection - Full params:', params);
    
    const callSid = params.CallSid || params.callsid || 'unknown';
    const digits = params.digits || params.Digits || '';
    
    console.log(`Bot selection - CallSid: ${callSid}, Selected: ${digits}`);
    
    // Get or create session
    if (!callSessions[callSid]) {
      console.log(`Creating new session for call ${callSid}`);
      callSessions[callSid] = {
        callerNumber: params.CallFrom || params.From || 'unknown',
        stage: 'bot-selection',
        language: 'en',
        timestamp: new Date().toISOString()
      };
    }
    
    // Update session with bot selection
    const session = callSessions[callSid];
    session.stage = 'bot-selected';
    
    if (LEGAL_BOTS[digits]) {
      botName = LEGAL_BOTS[digits].name;
      console.log(`Selected bot: ${botName} for call ${callSid}`);
      
      // Store the selected bot
      session.selectedBot = LEGAL_BOTS[digits].name;
      
      // Now prompt to record a question instead of just playing an audio response
      const exotelResponse = `<?xml version="1.0" encoding="UTF-8"?>
<Response>
<Play>${process.env.AUDIO_BASE_URL}/please-ask-question.mp3</Play>
<Record action="${getBaseUrl(req)}/process-recording" maxLength="30" finishOnKey="#" playBeep="true"/>
</Response>`;
      
      res.set('Content-Type', 'text/xml');
      return res.send(exotelResponse);
    } else {
      console.log(`Invalid selection: ${digits} for call ${callSid}`);
      
      // Invalid selection response
      const exotelResponse = `<?xml version="1.0" encoding="UTF-8"?>
<Response>
<Play>${process.env.AUDIO_BASE_URL}/invalid-selection.mp3</Play>
<Gather action="${getBaseUrl(req)}/bot-selection" numDigits="1"/>
</Response>`;
      
      res.set('Content-Type', 'text/xml');
      return res.send(exotelResponse);
    }
  } catch (error) {
    console.error('Error in bot selection:', error);
    const errorResponse = `<?xml version="1.0" encoding="UTF-8"?>
<Response>
<Play>${process.env.AUDIO_BASE_URL}/error.mp3</Play>
<Hangup/>
</Response>`;
    res.set('Content-Type', 'text/xml');
    return res.send(errorResponse);
  }
});

// NEW ENDPOINT: Process recording and convert speech to text
app.all('/process-recording', async (req, res) => {
  try {
    const params = {...req.query, ...req.body};
    
    const callSid = params.CallSid || params.callsid || 'unknown';
    const recordingUrl = params.RecordingUrl || '';
    const recordingDuration = params.RecordingDuration || '0';
    
    console.log(`Processing recording for call ${callSid}`);
    console.log(`Recording URL: ${recordingUrl}`);
    console.log(`Recording Duration: ${recordingDuration} seconds`);
    
    if (!callSessions[callSid]) {
      console.log(`No session found for call ${callSid}`);
      const errorResponse = `<?xml version="1.0" encoding="UTF-8"?>
<Response>
<Play>${process.env.AUDIO_BASE_URL}/error.mp3</Play>
<Hangup/>
</Response>`;
      res.set('Content-Type', 'text/xml');
      return res.send(errorResponse);
    }
    
    const session = callSessions[callSid];
    
    // Check if we have a recording URL and valid bot selection
    if (!recordingUrl || !session.selectedBot) {
      console.log('Missing recording URL or no bot selected');
      const errorResponse = `<?xml version="1.0" encoding="UTF-8"?>
<Response>
<Play>${process.env.AUDIO_BASE_URL}/error.mp3</Play>
<Play>${process.env.AUDIO_BASE_URL}/prompt-after-answer.mp3</Play>
<Gather action="${getBaseUrl(req)}/after-answer" numDigits="1"/>
</Response>`;
      res.set('Content-Type', 'text/xml');
      return res.send(errorResponse);
    }
    
    // Process recording - convert speech to text
    let transcriptionText = '';
    let botResponse = '';
    
    try {
      // Option 1: Use your existing RAG API's speech-to-text endpoint
      console.log('Sending recording to speech-to-text service...');
      const transcriptionResponse = await axios.post(`${RAG_API_URL}/speech-to-text`, {
        audio_url: recordingUrl,
        language: session.language || 'english'
      }, { timeout: 15000 });
      
      transcriptionText = transcriptionResponse.data.text || '';
      console.log(`Transcription result: "${transcriptionText}"`);
      
      // Store the transcription in the session
      session.lastQuery = transcriptionText;
      
      // If we have text, query the bot with it
      if (transcriptionText) {
        console.log(`Querying bot "${session.selectedBot}" with: "${transcriptionText}"`);
        const botQueryResponse = await axios.post(`${RAG_API_URL}/bots/${encodeURIComponent(session.selectedBot)}/query`, {
          query: transcriptionText
        }, { timeout: 15000 });
        
        botResponse = botQueryResponse.data.answer || '';
        console.log(`Bot response: "${botResponse}"`);
        
        // Store the bot response in session
        session.lastResponse = botResponse;
      } else {
        console.log('Transcription resulted in empty text');
        botResponse = "I couldn't understand your question. Please try again.";
      }
    } catch (error) {
      console.error('Error processing speech to text:', error);
      botResponse = "Sorry, I encountered an error processing your question.";
    }
    
    // Now we need to convert the text response to audio
    // Since we're working with pre-recorded audio files, we'll fall back to a generic response
    // In a real system, you'd use a TTS service to generate audio from botResponse
    
    let responsePath = `${process.env.AUDIO_BASE_URL}/generic-response.mp3`;
    
    // For now, just use the bot number to select a pre-recorded response
    const botNumber = Object.keys(LEGAL_BOTS).find(key => LEGAL_BOTS[key].name === session.selectedBot);
    if (botNumber) {
      responsePath = `${process.env.AUDIO_BASE_URL}/bot-${botNumber}-response.mp3`;
    }
    
    const exotelResponse = `<?xml version="1.0" encoding="UTF-8"?>
<Response>
<Play>${responsePath}</Play>
<Play>${process.env.AUDIO_BASE_URL}/prompt-after-answer.mp3</Play>
<Gather action="${getBaseUrl(req)}/after-answer" numDigits="1"/>
</Response>`;
    
    res.set('Content-Type', 'text/xml');
    return res.send(exotelResponse);
  } catch (error) {
    console.error('Error in process-recording:', error);
    const errorResponse = `<?xml version="1.0" encoding="UTF-8"?>
<Response>
<Play>${process.env.AUDIO_BASE_URL}/error.mp3</Play>
<Hangup/>
</Response>`;
    res.set('Content-Type', 'text/xml');
    return res.send(errorResponse);
  }
});

// After answer endpoint - updated to use audio
app.all('/after-answer', (req, res) => {
  try {
    const params = {...req.query, ...req.body};
    const callSid = params.CallSid || params.callsid || 'unknown';
    const digits = params.digits || params.Digits || '';
    
    console.log(`After answer - CallSid: ${callSid}, Choice: ${digits}`);
    
    if (digits === '1') {
      // Another question - go back to bot selection
      const exotelResponse = `<?xml version="1.0" encoding="UTF-8"?>
<Response>
<Play>${process.env.AUDIO_BASE_URL}/select-another-domain.mp3</Play>
<Gather action="${getBaseUrl(req)}/bot-selection" numDigits="1"/>
</Response>`;
      res.set('Content-Type', 'text/xml');
      return res.send(exotelResponse);
    } else {
      // End call
      console.log(`Ending call ${callSid}`);
      delete callSessions[callSid];
      const exotelResponse = `<?xml version="1.0" encoding="UTF-8"?>
<Response>
<Play>${process.env.AUDIO_BASE_URL}/goodbye.mp3</Play>
<Hangup/>
</Response>`;
      res.set('Content-Type', 'text/xml');
      return res.send(exotelResponse);
    }
  } catch (error) {
    console.error('Error in after-answer:', error);
    const errorResponse = `<?xml version="1.0" encoding="UTF-8"?>
<Response>
<Play>${process.env.AUDIO_BASE_URL}/thank-you.mp3</Play>
<Hangup/>
</Response>`;
    res.set('Content-Type', 'text/xml');
    return res.send(errorResponse);
  }
});

// Add debug test endpoint with minimal XML
app.get('/test-exotel', (req, res) => {
  const exotelResponse = `<?xml version="1.0" encoding="UTF-8"?><Response><Say voice="female">Test response</Say><Hangup/></Response>`;
  res.set('Content-Type', 'text/xml');
  return res.send(exotelResponse);
});

// Healthcheck endpoint
app.get('/health', (req, res) => {
  res.json({
    status: 'healthy',
    uptime: process.uptime(),
    activeCalls: Object.keys(callSessions).length,
    timestamp: new Date().toISOString()
  });
});

// Debug endpoint to view active sessions
app.get('/debug/sessions', (req, res) => {
  res.json({
    sessionCount: Object.keys(callSessions).length,
    sessions: callSessions
  });
});

// Start server
app.listen(PORT, () => {
  console.log(`Exotel IVR server running on http://localhost:${PORT}`);
  console.log(`Make sure Exotel Passthru applet is configured to call: https://your-ngrok-url/`);
  console.log('Current active sessions:', Object.keys(callSessions).length);
});
